staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go

```diff
@@ -17,330 +17,247 @@ limitations under the License.
-	GetAPIServiceNames() []string
+	AddAPIService(apiService *v1.APIService, handler http.Handler) error
+	UpdateAPIService(apiServiceName string) error
+	RemoveAPIService(apiServiceName string) error
 }
 
 const (
 	aggregatorUser                = "system:aggregator"
-	specDownloadTimeout           = 60 * time.Second
-	localDelegateChainNamePrefix  = "k8s_internal_local_delegation_chain_"
-	localDelegateChainNamePattern = localDelegateChainNamePrefix + "%010d"
+	specDownloadTimeout           = time.Minute
+	localDelegateChainNamePattern = "k8s_internal_local_delegation_chain_%010d"
 
 	// A randomly generated UUID to differentiate local and remote eTags.
 	locallyGeneratedEtagPrefix = "\"6E8F849B434D4B98A569B9D7718876E9-"
 )
 
-// IsLocalAPIService returns true for local specs from delegates.
-func IsLocalAPIService(apiServiceName string) bool {
-	return strings.HasPrefix(apiServiceName, localDelegateChainNamePrefix)
+// openAPISpecInfo is used to store OpenAPI spec with its priority.
+// It can be used to sort specs with their priorities.
+type openAPISpecInfo struct {
+	apiService v1.APIService
+	// This is the cache used to hold the specs fresh from the apiservices
+	spec cached.Replaceable[*spec.Swagger]
+
+	// The downloader is used only for non-local apiservices to
+	// re-update the spec every so often.
+	downloader cached.Data[*spec.Swagger]
 }
 
-// GetAPIServiceNames returns the names of APIServices recorded in specAggregator.openAPISpecs.
-// We use this function to pass the names of local APIServices to the controller in this package,
-// so that the controller can periodically sync the OpenAPI spec from delegation API servers.
-func (s *specAggregator) GetAPIServiceNames() []string {
-	names := make([]string, 0, len(s.openAPISpecs))
-	for key := range s.openAPISpecs {
-		names = append(names, key)
+type specAggregator struct {
+	mutex sync.Mutex
+
+	// Map of API Services' OpenAPI specs by their name
+	openAPISpecs map[string]*openAPISpecInfo
+
+	// provided for dynamic OpenAPI spec
+	openAPIVersionedService *handler.OpenAPIService
+
+	downloader *Downloader
+}
+
+func buildAndRegisterSpecAggregatorForLocalServices(downloader *Downloader, aggregatorSpec *spec.Swagger, delegationHandlers []http.Handler, pathHandler common.PathHandler) *specAggregator {
+	s := &specAggregator{
+		downloader:   downloader,
+		openAPISpecs: map[string]*openAPISpecInfo{},
+	}
+
+	cachedSpec := cached.NewStaticSource(func() cached.Result[*spec.Swagger] {
+		return cached.NewResultOK(aggregatorSpec, "never-changes")
+	})
+
+	s.addLocalSpec(fmt.Sprintf(localDelegateChainNamePattern, 0), cachedSpec)
+
+	for i, handler := range delegationHandlers {
+		name := fmt.Sprintf(localDelegateChainNamePattern, i+1)
+
+		spec := NewCacheableDownloader(downloader, handler)
+		spec = newNameDecoratedErrorCache(name, spec)
+		s.addLocalSpec(name, spec)
 	}
-	return names
+	s.openAPIVersionedService = handler.NewOpenAPIServiceLazy(s.buildMergeSpecLocked())
```

Comment:
Empty line before. Use whitespaces intentionally where they help structuring the code. Here, one part is spec building, the next is spec registration. Deserves an empty line. In contrast, some of those above don't make sense.

Reply:
Thanks for pointing that out. Updated the empty lines above as well.